resolver/dns/dns_resolver_test.go

```diff
@@ -1126,3 +1142,80 @@ func TestCustomAuthority(t *testing.T) {
 		}
 	}
 }
+
+// TestRateLimitedResolve exercises the rate limit enforced on re-resolution
+// requests. It sets the re-resolution rate to a small value and repeatedly
+// calls ResolveNow() and ensures only the expected number of resolution
+// requests are made.
+func TestRateLimitedResolve(t *testing.T) {
+	defer leakcheck.Check(t)
+
+	dc := replaceDNSResRate(250 * time.Millisecond)
+	defer dc()
+
+	// Create a new testResolver{} for this test because we want the exact count
+	// of the number of times the resolver was invoked.
+	nc := replaceNetFunc()
+	defer nc()
+
+	target := "foo.bar.com"
+	b := NewBuilder()
+	cc := &testClientConn{target: target}
+	r, err := b.Build(resolver.Target{Endpoint: target}, cc, resolver.BuildOption{})
+	if err != nil {
+		t.Fatalf("resolver.Build() returned error: %v\n", err)
+	}
+	defer r.Close()
+
+	dnsR, ok := r.(*dnsResolver)
+	if !ok {
+		t.Fatalf("resolver.Build() returned unexpected type: %T\n", dnsR)
+	}
+	tr, ok := dnsR.resolver.(*testResolver)
+	if !ok {
+		t.Fatalf("delegate resolver returned unexpected type: %T\n", tr)
+	}
+
+	// We set the re-resolution rate to 250ms at the start of this test. We
+	// expect the first call to ResolveNow() to make a resolution request
+	// immediately (T0), and one amongst the other queued up calls to make
+	// another resolution request at (T0+250ms). We wait till a little longer to
+	// make sure that more resolution requests are not made.
+	start := time.Now()
+	end := start.Add(100 * time.Millisecond)
+	calls := 0
+	for time.Now().Before(end) {
+		r.ResolveNow(resolver.ResolveNowOption{})
+		calls++
+		time.Sleep(1 * time.Millisecond)
+	}
+	if calls <= 2 {
+		t.Fatalf("expected to make more than two calls to ResolveNow, made only %d", calls)
+	}
+	time.Sleep(1 * time.Second)
```

Comment:
I'm not fond of sleeps in tests when they can be avoided.

Maybe we can instrument the dnsResolver a bit more so that we can inject signals through channels in tests instead of relying on time?

Reply:
We could essentially let the first resolution request happening as part of d.t.C run its course, and then start the `ResolveNow()` storm, and expect exactly two requests. Does that seem reasonable to you? But we would still need a sleep, before the `ResolveNow()` storm, to make sure that the first resolution happened and we got past the sleep for `minDNSRate`.

Any other ideas?

Reply:
A couple ideas:

1. Factor the sleeping part of the code into a function that we replace when testing with something that blocks on a channel instead.  (I don't really like stubbing out too much real code when testing, but it could work.)

2. Instead of sleeping and checking what happened when we wake up, call `ResolveNow` frequently in a loop in one goroutine (with ~1ms sleep between calls).  The main Test goroutine waits until it sees X resolutions happen (this can be signaled by the test `ClientConn` closing a channel).  Then it checks the time elapsed since start and ensures it's at least `X*minDNSResRate`.  Then we can set poll interval to something small like 10ms, and X to something like 5.  That should not be flaky and will complete quickly.

I think I would prefer #2.  WDYT?

Reply:
Done.
I restructured the test a little so that the test could be notified of resolver events, and therefore doesn't have to sleep anymore. I made sure it is not flaky with -count=100.